src/main/java/com/bookstore/service/BookServiceImpl.java

```diff
@@ -20,10 +22,17 @@ public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
+    private final CategoryRepository categoryRepository;
 
     @Override
     public BookDto save(CreateBookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
+        requestDto.getCategoryIds()
+                .stream()
+                .map(categoryRepository::findById)
+                .map(c -> c.orElseThrow(() ->
+                        new EntityNotFoundException("can't find category: " + c)))
+                .forEach(category -> category.addBook(book));
```

Comment:
Let the mapper do it
And without fetching from the DB - just Category with id field